crates/networking/p2p/rlpx/l2/l2_connection.rs

```diff
@@ -0,0 +1,934 @@
+                ))?;
+            let new_block_header = established.storage.get_block_header(block_number)?.ok_or(
+                RLPxError::InternalError(
+                    "Block header not found after querying for the block number".to_owned(),
+                ),
+            )?;
+            let new_block = Block {
+                header: new_block_header,
+                body: new_block_body,
+            };
+            let (signature, recovery_id) = if let Some(recovered_sig) = l2_state
+                .store_rollup
+                .get_signature_by_block(new_block.hash())
+                .await?
+            {
+                let mut signature = [0u8; 64];
+                let mut recovery_id = [0u8; 4];
+                signature.copy_from_slice(&recovered_sig[..64]);
+                recovery_id.copy_from_slice(&recovered_sig[64..68]);
+                (signature, recovery_id)
+            } else {
+                let (recovery_id, signature) = secp256k1::SECP256K1
+                    .sign_ecdsa_recoverable(
+                        &SecpMessage::from_digest(new_block.hash().to_fixed_bytes()),
+                        &l2_state.committer_key,
+                    )
+                    .serialize_compact();
+                let recovery_id: [u8; 4] = recovery_id.to_i32().to_be_bytes();
+                (signature, recovery_id)
+            };
+            NewBlock {
+                block: new_block.into(),
+                signature,
+                recovery_id,
+            }
+        };
+
+        send(established, new_block_msg.into()).await?;
+        established
+            .l2_state
+            .connection_state_mut()?
+            .latest_block_sent = block_number;
+    }
+
+    Ok(())
+}
+
+async fn should_process_new_block(
+    established: &mut Established,
+    msg: &NewBlock,
+) -> Result<bool, RLPxError> {
+    let l2_state = established.l2_state.connection_state_mut()?;
+    if !established.blockchain.is_synced() {
+        debug!("Not processing new block, blockchain is not synced");
+        return Ok(false);
+    }
+    if l2_state.latest_block_added >= msg.block.header.number
+        || l2_state
+            .blocks_on_queue
+            .contains_key(&msg.block.header.number)
+    {
+        debug!(
+            "Block {} received by peer already stored, ignoring it",
+            msg.block.header.number
+        );
+        return Ok(false);
+    }
+
+    let block_hash = msg.block.hash();
+
+    let recovered_lead_sequencer = get_pub_key(
+        msg.recovery_id,
+        &msg.signature,
+        *block_hash.as_fixed_bytes(),
+    )
+    .map_err(|e| {
+        log_peer_error(
+            &established.node,
+            &format!("Failed to recover lead sequencer: {e}"),
+        );
+        RLPxError::CryptographyError(e.to_string())
+    })?;
+
+    if !validate_signature(recovered_lead_sequencer) {
+        return Ok(false);
+    }
+    let mut signature = [0u8; 68];
```

Comment:
Why 68 bytes for the signature? Oh I think we talked about this async. Let's file an issue to reduce the size to 65 (64 bytes for the signature and 1 byte for the recovery ID).